### PR TITLE
Task #34174 send email to user after 2h if user created exam but don'…

### PR DESCRIPTION
**Pull request by phuongbk**
…t do it

### CODE REVIEW

File: app/controllers/exams_controller.rb [hunk 1]
```diff
@@ -12,6 +12,7 @@ def new
   def create
     if @exam.save
       redirect_to root_url , success: t("messages.exams.created")
+      ExamWorker.perform_async(@exam.id)
```
Comment (yuyuvn) [inline app/controllers/exams_controller.rb]: remove ()

File: app/workers/exam_worker.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class ExemWorker
+  include Sidekiq::Worker
+
+  def perform_at(exam_id)
+    exam = Exam.find(@exam_id)
```
Comment (yuyuvn) [inline app/workers/exam_worker.rb]: remove ()

**Comment by vuducdat** [comment]
I same Duc's Comment. LGTM

File: app/workers/exam_worker.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class ExemWorker
+  include Sidekiq::Worker
+
+  def perform_at(exam_id)
```
Comment (vanlinha3k35pbc) [inline app/workers/exam_worker.rb]: remove ()

File: app/workers/exam_worker.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+    exam = Exam.find @exam_id
```
Comment (vanlinha3k35pbc) [inline app/workers/exam_worker.rb]: exam_id

File: app/workers/exam_worker.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    if !started?
```
Comment (vanlinha3k35pbc) [inline app/workers/exam_worker.rb]: unless exam.started?

File: app/controllers/exams_controller.rb [hunk 6]
```diff
@@ -12,6 +12,7 @@ def new
   def create
     if @exam.save
       redirect_to root_url , success: t("messages.exams.created")
+      ExamWorker.perform_async @exam
```
Comment (vanlinha3k35pbc) [inline app/controllers/exams_controller.rb]: nên truyền tham số là id
https://github.com/mperham/sidekiq/wiki/Best-Practices

File: app/workers/exam_worker.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class ExemWorker
+  include Sidekiq::Worker
+
+  def perform_at exam
+    unless exam.started?
+      ExamMailer.delay_for(Settings.delay_time.minutes).remine_user(exam).deliver
```
Comment (vanlinha3k35pbc) [inline app/workers/exam_worker.rb]: sao lại delay ?

File: app/workers/exam_worker.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class ExemWorker
+  include Sidekiq::Worker
+
+  def perform_in(timestamp, exam_id)
```
Comment (vanlinha3k35pbc) [inline app/workers/exam_worker.rb]: perform exam_id

**Comment by vanlinha3k35pbc** [comment]
OK

File: app/workers/exam_worker.rb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+class ExemWorker
+  include Sidekiq::Worker
+
+  def perform exam_id
+    exam = Exam.find exam_id
+    unless exam.started?
```
Comment (dieunb) [inline app/workers/exam_worker.rb]: one line of code

File: app/controllers/exams_controller.rb [hunk 10]
```diff
@@ -12,6 +12,7 @@ def new
   def create
     if @exam.save
       redirect_to root_url , success: t("messages.exams.created")
+      ExamWorker.perform_in Settings.delay_time.hours, @exam.id
```
Comment (dieunb) [inline app/controllers/exams_controller.rb]: move to call_back :D

File: app/models/exam.rb [hunk 11]
```diff
@@ -13,6 +13,8 @@ class Exam < ActiveRecord::Base
   before_save ->{self.correct_number = results.select{|result|
     result.is_correct?}.count}
 
+  after_save ->{ExamWorker.perform_in Settings.delay_time.hours, self.id}
```
Comment (yuyuvn) [inline app/models/exam.rb]: after_create

**Pull request merged**